docs/best-practices.md

```diff
@@ -0,0 +1,799 @@
+        Self {
+            data: data.into(),
+        }
+    }
+
+    pub fn get_data(self) -> Base64VecU8 {
+        self.data.into()
+    }
+}
+```
+
+## View vs Change method
+
+`near_sdk` assumes that the method is a `view` if it uses `&self` or `self` and method is `change` if it has `&mut self`.
+
+View methods don't save the contract STATE at the end of the method execution.
+However, a view method COULD modify contract STATE or persistent collection state in-memory, knowing that all changes will be discarded after the method returns.
+
+Change methods will automatically save the modified STATE at the end of the method execution. They can also modify the state in persistent collections.
+
+Note: Change methods will also check that the function call doesn't have attached deposit, unless the method is marked with the `#[payable]` macro.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    /// View method. Requires cloning the account id.
+    pub fn get_owner_id(&self) -> AccountId {
+        self.owner_id.clone()
+    }
+
+    /// View method. More efficient, but can't be reused internally, because it consumes self.
+    pub fn get_owner_id2(self) -> AccountId {
+        self.owner_id
+    }
+
+    /// Change method. Changes the state, and then saves the new state internally.
+    pub fn set_owner_id(&mut self, new_owner_id: AccountId) {
+        self.owner_id = new_owner_id;
+    }
+
+    /// View method that "modifies" state, for code structure or computational
+    /// efficiency reasons. Changes state in-memory, but does NOT save the new
+    /// state. If called internally by a change method, WILL result in updated
+    /// contract state.
+    pub fn update_stats(&self, account_id: AccountId, score: U64) -> Account {
+        let account = self.accounts.get(&account_id).expect("account not found");
+        account.total += score;
+        account
+    }
+}
+```
+
+For more information about `&self` versus `self` see the [rust book](https://doc.rust-lang.org/stable/book/ch05-03-method-syntax.html?highlight=capture%20self#defining-methods)
+
+## Payable methods
+
+To mark a change method as a payable, you need to add the `#[payable]` macro decorator. This will allow this change method
+to receive attached deposits. Otherwise, if a deposit is attached to a non-payable change method, the method will panic.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    #[payable]
+    pub fn take_my_money(&mut self) {
+        env::log_str("Thanks!");
+    }
+
+    pub fn do_not_take_my_money(&mut self) {
+        env::log_str("Thanks!");
+    }
+}
+```
+
+This is equivalent to:
+
+```rust
+#[near_bindgen]
+impl Contract {
+    pub fn take_my_money(&mut self) {
+        env::log_str("Thanks!");
+    }
+
+    pub fn do_not_take_my_money(&mut self) {
+        if near_sdk::env::attached_deposit() != 0 {
+            near_sdk::env::panic(b"Method do_not_take_my_money doesn't accept deposit");
+        }
+        env::log_str("Thanks!");
+    }
+}
+```
```

Comment:
already covered

Reply:
migrated the example here as it's useful: https://github.com/near/sdk-docs/pull/37/commits/742be704a481a0b74070dd6373ac4c4230043356